dbms/src/Core/OperatorSpillContext.h

```diff
@@ -56,14 +61,22 @@ class OperatorSpillContext
             return 0;
     }
     UInt64 getOperatorSpillThreshold() const { return operator_spill_threshold; }
-    void markSpill()
+    void markSpilled()
     {
-        SpillStatus init_value = SpillStatus::NOT_SPILL;
-        if (spill_status.compare_exchange_strong(init_value, SpillStatus::SPILL, std::memory_order_relaxed))
+        bool init_value = false;
+        if (is_spilled.compare_exchange_strong(init_value, true, std::memory_order_relaxed))
         {
             LOG_INFO(log, "Begin spill in {}", op_name);
         }
     }
-    bool isSpilled() const { return spill_status != SpillStatus::NOT_SPILL; }
+    bool isSpilled() const { return is_spilled; }
+    /// auto trigger spill means the operator will auto spill under the constraint of query/global level memory threshold,
+    /// so user does not need set operator_spill_threshold explicitly
+    virtual bool supportAutoTriggerSpill() const { return false; }
+    virtual Int64 triggerSpill(Int64 expected_released_memories) = 0;
 };
+
+using OperatorSpillContextPtr = std::shared_ptr<OperatorSpillContext>;
+using RegisterOperatorSpillContext = std::function<void(const OperatorSpillContextPtr & ptr)>;
+using AutoSpillTrigger = std::function<void()>;
```

Comment:
Looks like you could use a class instead of a lambda, which would reduce the calling overhead.

Reply:
OK, use a class instead of lambda here.